### ISSUE TITLE
machine needs to be able to re-detect hosts its made

### ISSUE THREAD

**Issue by SvenDowideit** [docker/machine#248]
this may speak against making the droplet/vm/etc name the same as the user-friendly machine name, as at this point, you could implement a `machine detect --driver=digitalocean` or `vmwarefusion` etc and have all the orphans re-added to my list of machines automatically.

tbh, this kind of recovery seems like an important thing, given that the cloud ones will cost money.

**Comment by ehazlett** [comment]
I like the idea, however as @sthulb mentioned, I'm not sure how we would handle the keys in some cases.  If you delete the .docker dir or hardware fail or different system, the private ssh keys won't be there as machine generates these per machine.

There is a similar idea in #180 for b2d as well.  Maybe we could at least show a list or something of what is there.

**Comment by SvenDowideit** [comment]
the ssh keys are a problem, as are the TLS host keys, but at least the user would know their being charged for it :)

**Comment by sthulb** [comment]
It depends, this is perhaps a really hard thing to check. 

Cloud accounts can be shared, so how do we check which instance belongs to whom?

**Comment by ehazlett** [comment]
Yeah I'm curious how we would filter non-machine instances as well.

**Comment by Tebro** [comment]
Could we modify the images to give out something specific on some kind of nmap like scan?

**Comment by nathanleclaire** [comment]
Ideally, machine should "tag" cloud instances that it creates with an identifier which indicates it was created by Docker machine.  I know Amazon supports such tagging functionality, but we may have to try and find a way to work around by setting other metadata such as name with a certain pattern to indicate that it was created by machine.

**Comment by sthulb** [comment]
We can perhaps discover them, but how do we recover keys?

**Comment by nathanleclaire** [comment]
> We can perhaps discover them, but how do we recover keys?

We wouldn't, unfortunately.  Assuming something drastic like deleting `.docker` has happened, the machines would be in a "Disaster" state or similar and user would be allowed to perform basic actions like `Remove` but they would not be able to do anything that requires keys.

I think it kind of hints at needing a better way to replicate all the things we need to be present in `.docker` so that you can manage your cluster from another computer (and do so securely), but I'd rather leave that for another issue.

**Comment by sthulb** [comment]
I like this idea, but we need to solve shared accounts. 

For example, two users of the same AWS account, both with machine instances. 

How do we deal with that?

**Comment by SvenDowideit** [comment]
I'd suggest solving the ones we can solve first - mind you, I also suspect that if we're sharing AWS accounts, we want to know about each other's machines, and will want to add some docker daemon labels to communicate.

**Comment by Tebro** [comment]
For key "recovery": How about adding an option to generate a new key that gets associated with a particular machine? The user would of course have to install this new key manually, but there really is no good way around that without compromising security.

**Comment by ehazlett** [comment]
+1 to @Tebro for using existing keys -- this gets a little tricky with the various providers but would be a great benefit i think.  it would also help in the "recovery".

**Comment by nathanleclaire** [comment]
Although with some providers you might be hosed, with an API as robust as, say, Amazon's you can rotate SSH keys so those might be able to "auto-recover".

**Comment by md5** [comment]
@nathanleclaire I'm not so sure about that:

> Amazon EC2 doesn't keep a copy of your private key; therefore, if you lose your private key, there is no way to recover it. If you lose the private key for an instance store-backed instance, you can't access the instance; you should terminate the instance and launch another instance using a new key pair. If you lose the private key for an EBS-backed instance, you can regain access to your instance. For more information, see [Connecting to Your Linux Instance if You Lose Your Private Key](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html#replacing-lost-key-pair).

The linked doc talks about reattaching the EBS root volume on a different instance to modify the keys in the authorized_keys file of the appropriate user before reattaching it to the original instance. I think this is out of scope for `machine` to do (at least beyond documentation).

**Comment by nathanleclaire** [comment]
Ah, yes, I mis-remembered.  Good callout @md5.  The key name is passed as a parameter when you run the instance, and it can't be modified once the instance has been created, so I think in that case we just have to bootstrap discovery well enough so that the user can at least delete the instance and associated resources using `docker-machine`.
